File: benchmarks/raytracer/raytracer-cpp/src/ray.cpp
#pragma once

#include <math.h>
#include "vector.cpp"
#include "intersection.cpp"
#include "scene.cpp"
#include "sphere.cpp"


/*
 * Class holding data about a ray
 */
class Ray {

	public:
		// the origin location of the ray
		Vector origin;
		// the direction of the ray
		Vector direction;

		static Ray create(const Vector *origin, const Vector *direction) {
			Ray ray;
			ray.origin = *origin;
			ray.direction = *direction;
			return ray;
		}

		/**
		 * Reflect this ray at the given position and normal. Return the result as a new ray.
		 */
		Ray reflect(const Vector *position, const Vector *normal) const {
			Vector reflected = Vector::reflectThrough(&direction, normal);
			return Ray::create(position, &reflected);
		}

		/*
		 * Test the given sphere for an intersection with this ray and return the intersection (or Intersection.NONE)
		 */
		Intersection intersectSphere(const Sphere *sphere) const {
			Vector oc = Vector::sub(&origin, &(sphere->center));
			float a = Vector::dot(&direction, &direction);
			float b = 2.0f * Vector::dot(&oc, &direction);
			float c = Vector::dot(&oc, &oc) - sphere->radius*sphere->radius;
			float discriminant = b*b - 4.0f*a*c;
			if(discriminant > 0.0f) {
				float distance = (-b - sqrt(discriminant)) / (2.0f*a);
				Vector position = Vector::add(&origin, &((const Vector &) Vector::scale(&direction, distance)));
				Vector normal = Vector::unitVector(&((const Vector &) Vector::sub(&position, &(sphere->center))));
				return Intersection::create(true, distance, &position, &normal, sphere);
			} else {
				return Intersection::NONE;
			}
		}

		/*
		 * Test for intersections between this ray and an object in the given scene. Return the closest intersection or Intersection.NONE.
		 */
		Intersection intersectScene(const Scene *scene) const {
			Intersection closest = Intersection::NONE;
			for(int i=0; i<scene->objects.size(); i++) {
				const Sphere object = scene->objects[i];
				const Intersection intersection = intersectSphere(&object);
				if(intersection.hitSth && intersection.distance < closest.distance) {
					closest = intersection;
				}
			}
			return closest;
		}

};

File: benchmarks/raytracer/raytracer-js/src/ray.js
import { Vector } from './vector.js';
import { Intersection } from './intersection.js';
import { Scene } from './scene.js';
import { Sphere } from './sphere.js';

/*
 * Class holding data about a ray
 */
export class Ray {

	constructor(origin, direction) {
		// the origin location of the ray (vector)
		this.origin = origin;
		// the direction of the ray (vector)
		this.direction = direction;
	} 


	/**
	 * Reflect this ray at the given position and normal. Return the result as a new ray.
	 */
	reflect(position, normal) {
		return new Ray(position, Vector.reflectThrough(this.direction, normal));
	}

	/**
	 * Test for intersections between this ray and an object in the given scene. Return the closest intersection or Intersection.NONE.
	 */
	intersectScene(scene) {
		let closest = Intersection.NONE;
		for(let i=0; i<scene.objects.length; i++) {
			const object = scene.objects[i];
			const intersection = this.intersectSphere(object);
			if(intersection.hitSth && intersection.distance < closest.distance) {
				closest = intersection;
			}
		}
		return closest;
	}

	/**
	 * Test the given sphere for an intersection with this ray and return the intersection (or Intersection.NONE)
	 */
	intersectSphere(sphere) {
		const oc = Vector.sub(this.origin, sphere.center);
		const a = Vector.dot(this.direction, this.direction);
		const b = 2.0 * Vector.dot(oc, this.direction);
		const c = Vector.dot(oc,oc) - sphere.radius*sphere.radius;
		const discriminant = b*b - 4*a*c;
		if(discriminant > 0) {
			const distance = (-b - Math.sqrt(discriminant)) / (2.0*a);
			const position = Vector.add(this.origin, Vector.scale(this.direction, distance));
			const normal = Vector.unitVector(Vector.sub(position, sphere.center));
			return new Intersection(true, distance, position, normal, sphere);
		} else {
			return Intersection.NONE;
		}
	}

}
